### PR TITLE
4.3 Branch z2ft Integration

### PR DESCRIPTION
**Pull request by TheRealQuantam**
These are the initial set of changes for z2ft integration into the 4.3 branch. There are still a couple minor bugs to fix, but they're not something that need block getting it into people's hands.

Right now the primary purpose of the PR is to start a code review. It is NOT ready for merge quite yet.

The changes can be broken down into a few categories:
- Increase PRG-ROM size to 256 KB for music and anything else it might be useful for in the randomizer (e.g. rooms)
- Add constants for ROM sizes and locations and change all references to CHR-ROM to use the constants
- Change the HUD from spite 0 to IRQ-based to free up CPU time for music playback. As there is a policy that the randomizer should not lag less than the vanilla game, this change is only applied when music is randomized.
- Apply z2ft when music is randomized
- Integrate FtRandoLib (https://github.com/TheRealQuantam/FtRandoLib)
- Music randomization via FtRandoLib
- UI changes: 2 new customization booleans to randomize music (default off) and disable unsafe music (default on, though it doesn't do anything without randomize music)

For testing:
- Put FtRandoLib in the Z2Randomizer folder
- Create a Music directory in the same directory as the EXE and put music libraries (e.g. on Discord) in it

### CODE REVIEW

File: Assembler/Init.s [hunk 1]
```diff
@@ -61,14 +61,30 @@ UPDATE_REFS target @ refs
 .segment "PRG5"   :bank $05 :size $4000 :mem $8000 :off $14010
 .segment "PRG6"   :bank $06 :size $4000 :mem $8000 :off $18010
 .segment "PRG7"   :bank $07 :size $4000 :mem $c000 :off $1c010
-.segment "CHR"    :size $20000 :off $20010 :out
+.segment "PRG10"   :bank $10 :size $2000 :mem $8000 :off $20010
```
Comment (TheRealQuantam) [inline Assembler/Init.s:64]: New banks. Do avoid having to make a bunch of changes, PRG0-PRG7 are left as is as 4K banks.

File: Assembler/js65/macpack/common.js [hunk 2]
```diff
@@ -1,4 +1,75 @@
 export const text = `
+;;; Standard register names
```
Comment (TheRealQuantam) [inline Assembler/js65/macpack/common.js:2]: Add standard register names so the codebase can get rid of explicit register numbers.

File: RandomizerCore/Hyrule.cs [hunk 3]
```diff
@@ -212,8 +213,11 @@ public Hyrule(RandomizerConfiguration config, BackgroundWorker worker, bool save
 
         bool raftIsRequired = IsRaftAlwaysRequired(props);
         bool passedValidation = false;
+        HashSet<int> freeBanks = new();
         while (palaces == null || palaces.Count != 7 || passedValidation == false)
         {
+            freeBanks = new(ROM.FreeRomBanks);
```
Comment (TheRealQuantam) [inline RandomizerCore/Hyrule.cs:219]: Currently not used but defined here so future changes can use some of the banks for additional rooms.

File: RandomizerCore/IpsPatcher.cs [hunk 4]
```diff
@@ -15,17 +15,10 @@ namespace Z2Randomizer.Core;
 /// </summary>
 internal class IpsPatcher
 {
-    public void Patch(byte[] romData, string patchName)
+    public void Patch(byte[] romData, byte[] ipsbyte)
```
Comment (TheRealQuantam) [inline RandomizerCore/IpsPatcher.cs:18]: Just add a version that doesn't require loading from disk.

File: RandomizerCore/Music.cs [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+    Encounter,
+    Cave,
+    Palace,
+    GreatPalace,
+    Boss,
+    LastBoss,
+    Credits,
+}
+
+internal class RomAccessAdapter : IRomAccess
```
Comment (TheRealQuantam) [inline RandomizerCore/Music.cs:34]: Required by FtRandoLib.

File: RandomizerCore/Music.cs [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+
+    public LoggerAdapter(StringBuilder sb) => this.sb = sb;
+
+    protected override void InternalWrite(string? message) => sb.Append(message);
+    protected override void InternalWriteLine(string? message) => sb.AppendLine(message);
+}*/
+
+internal class Z2Importer : Importer
+{
+    protected override int BankSize => 0x2000;
```
Comment (TheRealQuantam) [inline RandomizerCore/Music.cs:70]: Most of the music-related magic numbers.

**Comment by Ellendar** [comment]
We decided to do 4.3.13 to move forward some bugfixes (and disable L+R) so I'm going to push this along with.

**Pull request merged**